### ISSUE TITLE
Please specify Gil status in Readme.md

### ISSUE THREAD

**Issue by ytrezq** [tonybaloney/Pyjion#115]
It s well known IronPytthon and Jython leverage the power of their virtual machine backends in order to allow cpu intensive task requiring latency less shared memory parrallel python workloads. Is it the same with this implementation?

**Comment by tonybaloney** [comment]
There isn't a simple answer to this, but I'll try my best

IronPython and Jython (see README) are top-to-bottom Python interpreters. This is not. Pyjion is an extension to CPython that you load as a Python module.
It replaces the "frame evaluation" components, i.e. `EvalFrameDefault` in `ceval.c` with a replacement that compiles to machine code. The machine code is stored in memory and executed when that Python function is called.

It uses the CPython C-API to maintain 100% compatibility with any application that works in Python. It also defaults back into Cpython for things it doesn't support (like async methods).

It doesn't have a GIL in the same way that CPython does, because it doesn't have an interpreter loop. Instead of reading byte codes and interpreting them at runtime, it compiles them into sequential machine-code instructions.

However, when it calls another method, it doesn't know if Pyjion or CPython will execute that function, so it acquires the GIL. (That is a compilation switch if you want to try without it)

As for the overheads, I don't know what the overheads are like for the other projects. If you run a function once, and only once, it will be slower. JIT's are faster for "hot" code which is run many times. 

You can change the level at which the JIT kicks in with `set_threshold()` https://pyjion.readthedocs.io/en/latest/api.html#pyjion.set_threshold by default, it is 0.

**Comment by ytrezq** [comment]
What I meaning is it the code asked to be gitted is not using the gil. I m meaning would only mostly one thread run at time when using cpu bound threading with pure Pyjion?

**Comment by ytrezq** [comment]
@tonybaloney does the code running under the jit compiler is using the gil within itself? Or is it possible to perform cpu heavy workloads in parallel using threads if they only run pure jit code?

**Comment by tonybaloney** [comment]
If you want to accomplish that, you should use CPython 3.9's subinterpreters module. 

Pyjion does not try and address the complexities of parallel computation.

I've written about the subinterpreters module (before it was merged), if you're interested https://tonybaloney.github.io/posts/has-the-python-gil-been-slain.html

**Comment by ytrezq** [comment]
@tonybaloney no, many if not most C python modules don t implement `__reduce__()` https://github.com/Z3Prover/z3/issues/4935#issuecomment-755181068.

So in my case even if one thread at time access z3, I have to leave my cpu intensive work running during weeks single threaded leaving a gil free implementation (at least when not using code of third party C python modules).

It would be nice to have at lease an unsafe option for disabling the gil like with Numba (though I can t use Numba because many function have variables working on different object types using the same variables).

**Comment by tonybaloney** [comment]
You could comment out https://github.com/tonybaloney/Pyjion/blob/master/CMakeLists.txt#L139 and then it won't acquire the GIL when making function calls. For any other operation it doesn't acquire the GIL anyway.

Pyjion is still highly experimental, so I think you'll likely just crash Python at the moment. I haven't tested threading, there are many outstanding critical bugs which cause segmentation faults. Maybe in a few months it'd be worth trying.

**Comment by ytrezq** [comment]
@tonybaloney I might say a mistake, but that sound acheivable by making generated code thread safe, Numba can work safely without the Gil at all (if you make sure no Python code is called and that you specify the C like type of everything).
